charms/istio-pilot/src/resources_handler.py

```diff
@@ -0,0 +1,198 @@
+#! /usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Resources handling library using Lightkube."""
+import logging
+from typing import Any, Tuple, TextIO, Union, Iterable
+
+import yaml
+from jinja2 import Environment, FileSystemLoader
+import lightkube  # noqa F401  # Needed for patching in test_resources_handler.py
+from lightkube import Client, codecs
+from lightkube.core.exceptions import ApiError
+from lightkube.generic_resource import create_namespaced_resource, GenericNamespacedResource
+from lightkube.core.resource import Resource
+
+
+class ResourceHandler:
+    def __init__(self, app_name, model_name):
+        """A Lightkube API interface.
+
+        Args:
+           - app_name: name of the application
+           - model_name: name of the Juju model this charm is deployed to
+        """
+
+        self.app_name = app_name
+        self.model_name = model_name
+
+        self.log = logging.getLogger(__name__)
+
+        # Every lightkube API call will use the model name as the namespace by default
+        self.lightkube_client = Client(namespace=self.model_name, field_manager="lightkube")
+
+        self.env = Environment(loader=FileSystemLoader('src'))
+
+    def delete_object(
+        self, obj, namespace=None, ignore_not_found=False, ignore_unauthorized=False
+    ):
+        try:
+            self.lightkube_client.delete(type(obj), obj.metadata.name, namespace=namespace)
+        except ApiError as err:
+            self.log.exception("ApiError encountered while attempting to delete resource.")
+            if err.status.message is not None:
+                if "not found" in err.status.message and ignore_not_found:
+                    self.log.error(f"Ignoring not found error:\n{err.status.message}")
+                elif "(Unauthorized)" in err.status.message and ignore_unauthorized:
+                    # Ignore error from https://bugs.launchpad.net/juju/+bug/1941655
+                    self.log.error(f"Ignoring unauthorized error:\n{err.status.message}")
+                else:
+                    self.log.error(err.status.message)
+                    raise
+            else:
+                raise
+
+    def delete_existing_resource_objects(
+        self,
+        resource,
+        namespace=None,
+        ignore_not_found=False,
+        ignore_unauthorized=False,
+        labels=None,
+    ):
+        if labels is None:
+            labels = {}
+
+        for obj in self.lightkube_client.list(
+            resource,
+            labels={"app.juju.is/created-by": f"{self.app_name}"}.update(labels),
```

Comment:
I think this might be a remnant from when this was not a generic function.  Maybe remove this so it is truly generic?  

If you'd like a way of doing convenience labels across all calls of a ResourceHandler, you could always add an optional arg to `__init__` that way it is controllable.